test/widgets/action_sheet_test.dart

```diff
@@ -472,6 +476,14 @@ void main() {
       await tapCopyMessageTextButton(tester);
       await tester.pump(Duration.zero);
       check(await Clipboard.getData('text/plain')).isNotNull().text.equals('Hello world');
+
+      // regression check for #732
+      final snackbar = tester.widget<SnackBar>(find.byType(SnackBar));
```

Comment:
Oh interesting; thanks for noticing this. I'm not sure what could be going on there. I'm pretty sure I would have tried this when I suggested the regression test in https://github.com/zulip/zulip-flutter/pull/713#discussion_r1637100823 . One thing I'm curious about: does #732 still reproduce with the old code when you run the app?

I guess it's possible that some other change happened that ended up fixing it.

…aha! On version 0.0.16 on iOS, #732 actually doesn't reproduce anymore! So I guess that's what happened.

We still want to pass the correct `BuildContext`, though, since it's reasonable for it to be unmounted after the action sheet is popped, and it would be a bug to do things with it after it's unmounted. But this should go in its own commit, without the `Fixes` line. An earlier commit can add the regression check, and also leave out the `Fixes` line, and just say that the bug seems to have been fixed at some point after we discovered it.

Reply:
> One thing I'm curious about: does https://github.com/zulip/zulip-flutter/issues/732 still reproduce with the old code when you run the app?

Yeah, the bug is still there when we run the app on a real device (checked on my Android 12).

> On version 0.0.16 on iOS, https://github.com/zulip/zulip-flutter/issues/732 actually doesn't reproduce anymore!

It is reproduced on Android devices with `sdkInt <= 32`. If we put some logging when testing with a real device, the `BuildContext` is unmounted, but with tests, it is still mounted. So the bug still exists, it seems that the `BuildContext` popping mechanism is kind of weird in the testing environment.

Reply:
Ah OK; thanks for checking that.

Are you seeing the bug with both the "Copy message text" button and the "Copy link to message" button? One difference is that the former waits for a network request between the `.pop()` and the `copyWithPopup`. I wonder if the context is unmounted asynchronously, some duration after the `.pop()`, and that duration is racing with the network request. (Sometimes one finishes first, sometimes the other.)

I think this might be what's happening. On my iPhone, if I edit the "Copy message text" code to remove the `await fetchRawContentWithFeedback` (hard-coding `rawContent` to some string instead), then I replace it with `await Future.delayed(Duration(milliseconds: n));`, the bug reproduces when `n` is large (like 1000) but not when `n` is small (like 10). When it reproduces, `context.mounted` is false, as you've seen too; when it doesn't reproduce, it's true.

If this seems right to you, then I think the fix is to make the test simulate the network request taking much longer than the unmounting that it's racing against. I'm having trouble getting this to work properly, but it should be possible, and Greg may be able to help if you have trouble with it too.

Reply:
Yep, here's a (hacky) diff that succeeds in reproducing the issue after `git checkout @~~~ -- lib` to get the code from before the bugfix, but passes after the bugfix:
```diff
diff --git test/api/fake_api.dart test/api/fake_api.dart
index 1cbbffcdc..5e4411133 100644
--- test/api/fake_api.dart
+++ test/api/fake_api.dart
@@ -94,7 +94,7 @@ class FakeHttpClient extends http.BaseClient {
         return Future(() => throw exception);
       case _PreparedSuccess(:var bytes, :var httpStatus):
         final byteStream = http.ByteStream.fromBytes(bytes);
-        return Future(() => http.StreamedResponse(
+        return Future.delayed(Duration(milliseconds: 1000), () => http.StreamedResponse(
           byteStream, httpStatus, request: request));
     }
   }
diff --git test/widgets/action_sheet_test.dart test/widgets/action_sheet_test.dart
index db7852627..b54fa947d 100644
--- test/widgets/action_sheet_test.dart
+++ test/widgets/action_sheet_test.dart
@@ -411,7 +411,9 @@ void main() {
 
       prepareRawContentResponseSuccess(store, message: message, rawContent: 'Hello world');
       await tapCopyMessageTextButton(tester);
-      await tester.pump(Duration.zero);
+      for (int i = 0; i < 11; i++) {
+        await tester.pump(const Duration(milliseconds: 100));
+      }
       check(await Clipboard.getData('text/plain')).isNotNull().text.equals('Hello world');
 
       // regression check for #732
```

I'll see about wrapping the `fake_api.dart` side of that into an option you can pass to `FakeApiConnection.prepare`.

Then @sm-sayedi please experiment and try to simplify those magic numbers of 1000ms, 100ms, 11 iterations — don't need to spend a ton of time doing so, but if we can it'd be nice to reduce it to something that sheds a bit more light on just how this gets triggered.

Reply:
OK, I pushed a draft commit to the tip of this PR:
571df4b5e wip fake_api [nfc]: Add `delay` option to connection.prepare; TODO test

I'll need to write a test for that test code, and then can send it as a PR. In the meantime the only thing it's missing is such a test, so @sm-sayedi you can build atop that and plan to rebase atop the finished version once it lands.

With that commit in place, a somewhat less hacky diff that reproduces the issue is:
```diff
diff --git test/widgets/action_sheet_test.dart test/widgets/action_sheet_test.dart
index db7852627..ea91f632e 100644
--- test/widgets/action_sheet_test.dart
+++ test/widgets/action_sheet_test.dart
@@ -90,7 +90,7 @@ void main() {
     // Prepare fetch-raw-Markdown response
     // TODO: Message should really only differ from `message`
     //   in its content / content_type, not in `id` or anything else.
-    (store.connection as FakeApiConnection).prepare(json:
+    (store.connection as FakeApiConnection).prepare(delay: Duration(seconds: 1), json:
       GetMessageResult(message: eg.streamMessage(contentMarkdown: rawContent)).toJson());
   }
 
@@ -411,7 +411,9 @@ void main() {
 
       prepareRawContentResponseSuccess(store, message: message, rawContent: 'Hello world');
       await tapCopyMessageTextButton(tester);
-      await tester.pump(Duration.zero);
+      for (int i = 0; i < 11; i++) {
+        await tester.pump(const Duration(milliseconds: 100));
+      }
       check(await Clipboard.getData('text/plain')).isNotNull().text.equals('Hello world');
 
       // regression check for #732
```

Then the `prepareRawContentResponseSuccess` should take an optional duration to configure that behavior; I'll leave that for you to do as part of writing the test for the bugfix.

Reply:
Sent that `delay` feature as #770.

Reply:
> Are you seeing the bug with both the "Copy message text" button and the "Copy link to message" button?

Yeah, for the "Copy link to message" button the bug is present both in manual and widget testing. As you explained in [#713 comment](https://github.com/zulip/zulip-flutter/pull/713#discussion_r1657899344), `.pop()` takes a small amount of time to finish. When testing on my phone, it took around 20 milliseconds for the context to be unmounted.
For the "Copy message text" button the bug is only present in widget testing (which is resolved in the latest revision). In manual testing, `await fetchRawContentWithFeedback` gives the `BuildContext` enough time to be unmounted.